File: ModularMonolithVsMicroservice/common/src/jammer.cpp
#include "jammer.h"

#include <algorithm>

namespace pulsecore {

JammerDetector::JammerDetector(double power_threshold, double duty_cycle_threshold)
    : power_threshold_(power_threshold), duty_cycle_threshold_(duty_cycle_threshold) {}

void JammerDetector::Process(const pulse::IQBatch& batch, pulse::JamSummary* out) {
  const int n = batch.i_size();
  if (n > 0) {
    const double* i_arr = batch.i().data();
    const double* q_arr = batch.q().data();
    double power_sum = 0.0;
    uint64_t over_threshold = 0;
    for (int k = 0; k < n; ++k) {
      const double power = i_arr[k] * i_arr[k] + q_arr[k] * q_arr[k];
      power_sum += power;
      if (power >= power_threshold_) ++over_threshold;
    }

    const double mean_power = power_sum / static_cast<double>(n);
    const double duty_cycle =
        static_cast<double>(over_threshold) / static_cast<double>(n);

    ++batches_total_;
    if (duty_cycle >= duty_cycle_threshold_) ++batches_flagged_;
    max_duty_cycle_ = std::max(max_duty_cycle_, duty_cycle);
    max_mean_power_ = std::max(max_mean_power_, mean_power);
  }

  out->Clear();
  out->set_batches_total(batches_total_);
  out->set_batches_flagged(batches_flagged_);
  out->set_max_duty_cycle(max_duty_cycle_);
  out->set_max_mean_power(max_mean_power_);
}

}  // namespace pulsecore

File: ModularMonolithVsMicroservice/python/pulsecore/jammer.py
"""Duty-cycle-based jamming detector -- Python port of
common/src/jammer.cpp / common/include/jammer.h.
"""
from pulsecore import pulse_pb2


class JammerDetector:
    def __init__(self, power_threshold: float, duty_cycle_threshold: float):
        self._power_threshold = power_threshold
        self._duty_cycle_threshold = duty_cycle_threshold
        self._batches_total = 0
        self._batches_flagged = 0
        self._max_duty_cycle = 0.0
        self._max_mean_power = 0.0

    def process(self, batch: "pulse_pb2.IQBatch", out: "pulse_pb2.JamSummary") -> None:
        n = len(batch.i)
        if n > 0:
            power_threshold = self._power_threshold
            power_sum = 0.0
            over_threshold = 0
            # Packed columnar layout: zip over the two packed arrays,
            # no per-sample message access.
            for si, sq in zip(batch.i, batch.q):
                power = si * si + sq * sq
                power_sum += power
                if power >= power_threshold:
                    over_threshold += 1

            mean_power = power_sum / n
            duty_cycle = over_threshold / n

            self._batches_total += 1
            if duty_cycle >= self._duty_cycle_threshold:
                self._batches_flagged += 1
            if duty_cycle > self._max_duty_cycle:
                self._max_duty_cycle = duty_cycle
            if mean_power > self._max_mean_power:
                self._max_mean_power = mean_power

        out.Clear()
        out.batches_total = self._batches_total
        out.batches_flagged = self._batches_flagged
        out.max_duty_cycle = self._max_duty_cycle
        out.max_mean_power = self._max_mean_power
